lib/ansible/modules/clustering/mesos/marathon_app.py

```diff
@@ -0,0 +1,356 @@
+        else:
+            return {"message": "App '%s' already exist" % app_id}
+
+    def delete_app(self, data=None):
+        app_id = self.get_id(data)
+        ret, info = self.send_request('v2/apps' + app_id, method='DELETE', data=data)
+        self.changed = self.changed or bool(ret)
+        if ret:
+            self.wait_for_deployments('v2/apps' + app_id)
+        return info
+
+    def restart_app(self, data=None):
+        app_id = self.get_id(data)
+        ret, info = self.send_request('v2/apps' + app_id + '/restart', method='POST')
+        if not ret:
+            self.module.fail_json(msg=info)
+        self.changed = self.changed or bool(ret)
+        if ret:
+            self.wait_for_deployments('v2/apps' + app_id)
+        return info
+
+
+def main():
+    module_args = dict(
+        marathon_uri=dict(type='str', default='http://marathon.mesos:8080', aliases=['marathon_url', 'uri', 'url']),
+        state=dict(type='str', choices=['present', 'absent', 'update', 'replace', 'restart']),
+        url_username=dict(type='str', default=None, aliases=['username']),
+        url_password=dict(type='str', default=None, no_log=True, aliases=['password']),
+        file_reference=dict(required=False),
+        inline_data=dict(required=False),
+        sleep=dict(type='float', default=1.0),
+        wait=dict(type='bool', default=False),
+        force=dict(type='bool', default=False),
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True,
+        mutually_exclusive=(('file_reference', 'inline_data'),),
+        required_one_of=(('file_reference', 'inline_data'),),
+    )
+
+    if not has_lib_yaml:
+        module.fail_json(msg="missing python library: yaml")
+    if not has_lib_json:
+        module.fail_json(msg="missing python library: json")
+
+    if module.params['url_username'] is not None or module.params['url_password'] is not None:
+        module.params['force_basic_auth'] = True
+
+    inline_data = module.params['inline_data']
+    file_reference = module.params['file_reference']
+
+    if inline_data:
+        if not isinstance(inline_data, dict) and not isinstance(inline_data, list):
+            try:
+                data = json.loads(inline_data)
+            except:
+                data = yaml.safe_load(inline_data)
+        else:
+            data = inline_data
+    else:
+        try:
+            f = open(file_reference, "r")
+            data = json.load(f)
+            f.close()
+            if not data:
+                module.fail_json(msg="No valid data could be found.")
+        except:
+            module.fail_json(msg="The file '%s' was not found or contained invalid JSON data" % file_reference)
+
+    marathon = MarathonApp(module)
+
+    if module.params['state'] in ['absent', 'replace']:
+        body = marathon.delete_app(data)
+        marathon.results.append(body)
+    if module.params['state'] in ['present', 'update', 'replace']:
```

Comment:
Why so many states?

Reply:
I can merge state=present with state=update, at the moment they do the same thing: change elements if defined and preserve the existing ones. start=replace force to "reset" values:
```yaml
- name: Create App
  marathon_app:
    inline_data:
      id: /foo
      instances: 1
      cmd: sleep infinity
      cpus: 0.1
      mem: 16
    state: present
```
With sate=present (or state=update), unspecified attributes are preserved (cmd, mem, cpus):
```yaml
- name: Scale App
  marathon_app:
    inline_data:
      id: /foo
      instances: 10
    state: present
```

The state=reset guarantee that unspecified attributed will be deleted or set with default values (instances=1):
```yaml
- name: Reset App
  marathon_app:
    inline_data:
      id: /foo
      cmd: sleep infinity
      mem: 10
      cpus: 0.1
    state: reset
```